pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1706,37 +1706,53 @@ private CompletableFuture<Boolean> checkReplicationCluster(String remoteCluster)
-                    }
-                    return brokerService.pulsar().getPulsarResources().getClusterResources()
-                            .getClusterAsync(remoteCluster)
-                            .thenApply(clusterData ->
-                                    brokerService.getReplicationClient(remoteCluster, clusterData));
-                })
-                .thenAccept(replicationClient -> {
-                    if (replicationClient == null) {
-                        return;
+        CompletableFuture<Void> replicationStartFuture = new CompletableFuture<>();
+        AbstractReplicator.validatePartitionedTopicAsync(PersistentTopic.this.getName(), brokerService)
+            .thenCompose(__ -> brokerService.pulsar().getPulsarResources().getClusterResources()
+                    .getClusterAsync(remoteCluster)
+                    .thenApply(clusterData ->
+                            brokerService.getReplicationClient(remoteCluster, clusterData)))
+            .thenAccept(replicationClient -> {
+                Replicator replicator = replicators.computeIfAbsent(remoteCluster, r -> {
+                    try {
+                        return new GeoPersistentReplicator(PersistentTopic.this, cursor, localCluster,
+                                remoteCluster, brokerService, (PulsarClientImpl) replicationClient);
+                    } catch (PulsarServerException e) {
+                        log.error("[{}] Replicator startup failed {}", topic, remoteCluster, e);
                     }
-                    Replicator replicator = replicators.computeIfAbsent(remoteCluster, r -> {
-                        try {
-                            return new GeoPersistentReplicator(PersistentTopic.this, cursor, localCluster,
-                                    remoteCluster, brokerService, (PulsarClientImpl) replicationClient);
-                        } catch (PulsarServerException e) {
-                            log.error("[{}] Replicator startup failed {}", topic, remoteCluster, e);
+                    return null;
+                });
+                // clean up replicator if startup is failed
+                if (replicator == null) {
+                    replicators.removeNullValue(remoteCluster);
+                }
+            }).whenComplete((ignore, ex) -> {
+                if (ex == null){
+                    replicationStartFuture.complete(null);
+                } else {
+                    checkReplicationCluster(remoteCluster).thenCompose(clusterExists -> {
+                        if (!clusterExists) {
+                            log.warn("[{}] Start remove the replicator because the cluster '{}' does not exist",
+                                    topic, remoteCluster);
+                            replicationStartFuture.complete(null);
+                            return removeReplicator(remoteCluster).whenComplete((ignore2, remoteCursorEx) -> {
```

Comment:
why is the replicationStartFuture completed before removing the replicator?

Reply:
Already fixed. Thanks